src/core/frame.c

```diff
@@ -88,17 +88,19 @@ static void instrumentation_level_barrier(MVMThreadContext *tc, MVMStaticFrame *
         MVM_profile_ensure_uninstrumented(tc, static_frame);
 }
 
-/* Destroys a frame. */
+/* Called when the GC destroys a frame. Since the frame may have been alive as
+ * part of a continuation that was taken but never invoked, we should check
+ * things normally cleaned up on return don't need cleaning up also. */
 void MVM_frame_destroy(MVMThreadContext *tc, MVMFrame *frame) {
     if (frame->work) {
         MVM_args_proc_cleanup(tc, &frame->params);
         MVM_fixed_size_free(tc, tc->instance->fsa, frame->allocd_work,
             frame->work);
     }
-    if (frame->env)
-        MVM_fixed_size_free(tc, tc->instance->fsa, frame->allocd_env, frame->env);
     if (frame->continuation_tags)
         MVM_continuation_free_tags(tc, frame);
+    if (frame->env)
```

Comment:
could make the diff a tiny bit smaller (easier to digest?) by moving this back up before the continuation_free_tags part.